File: 2023/day-02/task1.mjs
import readline from 'node:readline';

const rl = readline.createInterface({ input: process.stdin });
const gameNumberRegex = /^Game (\d+?):/;
const redRegex = /(\d+?) red/g;
const greenRegex = /(\d+?) green/g;
const blueRegex = /(\d+?) blue/g;

const MAX_RED = 12;
const MAX_GREEN = 13;
const MAX_BLUE = 14;

let sum = 0;

for await (const line of rl) {
  const gameNumberMatch = line.match(gameNumberRegex);
  const gameNumber = gameNumberMatch && parseInt(gameNumberMatch[1], 10)

  if (isNaN(gameNumber)) {
    continue;
  }

  const redMatches = Array.from(line.matchAll(redRegex), m => parseInt(m, 10));
  const greenMatches = Array.from(line.matchAll(greenRegex), m => parseInt(m, 10));
  const blueMatches = Array.from(line.matchAll(blueRegex), m => parseInt(m, 10));

  const maxRed = Math.max(...redMatches)
  const maxGreen = Math.max(...greenMatches)
  const maxBlue = Math.max(...blueMatches)

  if (maxRed <= MAX_RED && maxGreen <= MAX_GREEN && maxBlue <= MAX_BLUE) {
    sum += gameNumber;
  }
}

console.log(sum);

File: 2023/day-02/task1.swift
import Foundation

let MAX_RED = 12
let MAX_GREEN = 13
let MAX_BLUE = 14

var total = 0

while let line = readLine() {
  let gameNumber = Int(line.split(separator: ":", maxSplits: 1)[0].filter(\.isNumber))!
  let splits = line.split(separator: " ")

  var maxRed = 0
  var maxGreen = 0
  var maxBlue = 0

  for i in 1...splits.count - 1 {
    let chunk = splits[i]

    if (chunk.starts(with: "red")) {
      maxRed = max(Int(splits[i - 1])!, maxRed)
    } else if (chunk.starts(with: "green")) {
      maxGreen = max(Int(splits[i - 1])!, maxGreen)
    } else if (chunk.starts(with: "blue")) {
      maxBlue = max(Int(splits[i - 1])!, maxBlue)
    }
  }

  if maxRed <= MAX_RED && maxGreen <= MAX_GREEN && maxBlue <= MAX_BLUE {
    total += gameNumber
  }
}

print(total)
